src/backend/distributed/transaction/transaction_management.c

```diff
@@ -281,12 +281,28 @@ CoordinatedTransactionCallback(XactEvent event, void *arg)
 
 		case XACT_EVENT_PREPARE:
 		{
+			/*
+			 * We should remove the intermediate results before unsetting the
+			 * distributed transaction id. That is necessary, otherwise Citus
+			 * would try to remove a non-existing folder and leak some of the
+			 * existing folders that are associated with distributed transaction
+			 * ids on the worker nodes.
+			 */
+			RemoveIntermediateResultsDirectory();
```

Comment:
Worth adding a note that this is called only on the worker, whereas most code in this function addresses the coordinator.